### ISSUE TITLE
FME_DEBBIE session merging

### ISSUE THREAD

**Issue by HenkMutsaerts** [ExploreASL/ExploreASL#1096]
Automatize the DEBBIE quantification by merging sessions as defined in a dataPar.json

### Tasks (based on what was discussed on the 11th of September)
What is done already in the main version:
- [x] Make file merging a separate function
- [x] HAD_8 processing -> before quant we save PWI4D.nii
        -> xASL_quant_Basil / line 141 : 
- [x] In dataPar.json, a list of pairs of sessions is provided -> x.modules.asl.SessionMergingList:[["ASL_1", "ASL_2"], ["ASL_3", "ASL_4"]] -> meaning ASL1+ASL2 need to be concatenated before quantification as well as ASL3+ASL4.
- [x] In xASL_quant_FSL, if SessonMergingList is recognized, it checks if the PWI4D of the "pair" session exists (based on what was provided by the SessionMergingList), and it concatenates the two.  'bMergingTxt' is then considered TRUE in order to merge the FSL_ModelOptions.txt inside xASL_sub_FSLOptions (function written inside xASL_quant_FSL)
- [x] The two txts are merged and all TIs, TEs, and tau's are added to FSL Options before quantification
- [x] T1b should be 1.7306679
- [x] T1 = 1.33
- [x] It gives labeling duration of ASL_2 for both ASL_1 and ASL_2 in FSL_ModulesOptions.txt - this affects both Tau and TI
- [x] Check if order of TE, TI, Tau, and NII is correct in the automatic concatenation.
- [x] Debug and clean
- [x] Create a new issue for using PLD-PWI4D from issue #997 to merge sequence parameters in the same function as NIIs are merged. #1543
- [x] Go through the entire xASL_wrp_QUantify -> xASL_quant_ASL_ -> xASL_quant_FSL and make sure that we use the PWI4D throughout without unnecessary loading and saving all the time. Just simply use the PWI4D which is made for this ;) We will probably need to do this after merging this and #997 issue. In #1543
- [x] in the merging function, we can create a merged PLD vector based on PLD_PWI4D. However, we only have access to PLD_PWI4D from the current session. And PLD_PWI4D from the previous sessions is not available in x-struct.
The solution is to do this in resample_ASL, we save together with PWI4D a simplified JSON containing the important PLD and other parameters and load this later. #1543
- [x] Save the merged file to a special file called PWI4D_Merged.nii. That way, we are still able to quantify PWI4D and PWI4D_Merged if we want to. We can’t move this to resampling, because M0_quant is called only after resampling and we might need this for the merging. We add sessions to the name PWI4D_Merged_1_2 and then delete this after using. #1543
- [x] quant_FSL is parsing PLDs etc and saves them to a TXT. Currently, it is also handling the parsing of DEBBIE parameters to the proper form. It also handles the concatenation of the parameters. The ideal solution is to only do PLD->FSL parsin, but move creating of PLD vectors (e.g. related to Hadamard) to the resampling function. The resampling function will create PLD_PWI4D vectors and we will only use in quant_FSL these vectors irrespective of if they come from DEBBIE or multi-PLD. Quant-FSL will thus not really deal with how PLD vectors are created. Currently quant_FSL also does the merging. In the future, the merging will go to Resample function as well. However, we are editing this concurently in #997 and #1096. So only after merging these two issues, we will fix the creation of PLD vectors. Note that this won’t still be super easy, as multi-TE needs to be handeled in a very special way and we can’t simply paste the vector of TEs. But that is doable automatically. So this part will have to be done in a follow-up issue. in #1543 
- [x] If only 1 M0 scan in session1, copy it to current session (needs to be done anyway). This is done for the last and all intermediary sessions. M0 has to be in the first session. So if we have ASL1+2+3 and M0_1, then this M0 is copied and processing re-run for ASL_2 and ASL_3 and of course ASL_1+2+3. #1543
- [x] We remove the the bSessionMerge parameter from xASL_quant_FSL and we provide path to the correct file. #1543
- [x] 	(improve later) Motion correction: motion correct HAD8 to HAD4. How? Copy HAD4 from session 1 and copy the first volume of HAD4. run an extra co-registration with the first volume of HAD4 and HAD8. Run motion corr of HAD8. Then apply this motion correction between HAD4 and HAD8 @jan-petr this will be probably done in a separate issue -> this issue doesn't exist yet, so has to be created and then added to the DEBBIE Master list. #1543
- [x] Can we have 3 sessions merged together for quantification? ASL_1 + ASL_2 + ASL_3. Test if ASL_1+2+4 works. Test if ASL_2+3+4 works. Test double pair ASL_1+2 and ASL_4+5. Test different sizes ASL_2+3 and ASL_1+4+5. #1543
- [x] Once we are done, we add SessionMergingList and bAutomaticMerging (or their final names) to the Documentation (inside #1096)

- [x] Change merging to concatenation (provided that we always do A A A A B B B B)
- [x] Create a new issue for `bAutomaticASLSessionMerging`
- [x] JAN: CHECK in the code #1543 comments by Henk, that are there for the next issue.
- [x] THIS BRANCH CANNOT BE MERGED TO THE DEVELOP BRANCH, IT WILL BE MERGED ONLY WHEN #997 AND #1543 ARE MERGED AS WELL

### Release notes

Automatic concatenation of sessions based on a provided list.

**Comment by BeatrizPadrela** [comment]
Okay, first step: change HAD8 to ASL_1 and HAD4 to ASL_2. In studyPar.json and sourcestructure.json:

![image](https://user-images.githubusercontent.com/73699072/232856451-a7447f70-121d-44d0-be19-9506545b4c7c.png)

**Comment by BeatrizPadrela** [comment]
Then, I'll try to check what happens if we have don't provide a session for M0 in the sourcestructure.json – and check if the pipeline can deal with it like it does with T1 and FLAIR scans.

**Comment by BeatrizPadrela** [comment]
We shall not forget to use 'xASL_exist' instead of 'exist' when checking if the PWI of the previous session exists
